fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java

```diff
@@ -176,6 +178,153 @@ private void handleRepayment(LoanTransaction loanTransaction, MonetaryCurrency c
+        List<LoanTransactionToRepaymentScheduleMapping> transactionMappings = new ArrayList<>();
+        LoanChargePaidBy loanChargePaidBy = loanTransaction.getLoanChargesPaid().stream().findFirst().get();
+        LoanCharge loanCharge = loanChargePaidBy.getLoanCharge();
+        Money amountToBePaid = Money.of(currency, loanTransaction.getAmount());
+        if (loanCharge.getAmountOutstanding(currency).isLessThan(amountToBePaid)) {
+            amountToBePaid = loanCharge.getAmountOutstanding(currency);
+        }
+
+        LocalDate startDate = loanTransaction.getLoan().getDisbursementDate();
+
+        Money unprocessed = loanTransaction.getAmount(currency);
+        for (final LoanRepaymentScheduleInstallment installment : installments) {
+            boolean isDue = installment.isFirstPeriod()
+                    ? loanCharge.isDueForCollectionFromIncludingAndUpToAndIncluding(startDate, installment.getDueDate())
+                    : loanCharge.isDueForCollectionFromAndUpToAndIncluding(startDate, installment.getDueDate());
+            if (isDue) {
+                Integer installmentNumber = installment.getInstallmentNumber();
+                Money paidAmount = loanCharge.updatePaidAmountBy(amountToBePaid, installmentNumber, zero);
+
+                LoanTransactionToRepaymentScheduleMapping loanTransactionToRepaymentScheduleMapping = getTransactionMapping(
+                        transactionMappings, loanTransaction, installment, currency);
+
+                if (loanTransaction.isPenaltyPayment()) {
+                    penaltyChargesPortion = installment.payPenaltyChargesComponent(loanTransaction.getTransactionDate(), paidAmount);
+                    loanTransaction.setLoanChargesPaid(Collections
+                            .singleton(new LoanChargePaidBy(loanTransaction, loanCharge, paidAmount.getAmount(), installmentNumber)));
+                    addToTransactionMapping(loanTransactionToRepaymentScheduleMapping, zero, zero, zero, penaltyChargesPortion);
+                } else {
+                    feeChargesPortion = installment.payFeeChargesComponent(loanTransaction.getTransactionDate(), paidAmount);
+                    loanTransaction.setLoanChargesPaid(Collections
+                            .singleton(new LoanChargePaidBy(loanTransaction, loanCharge, paidAmount.getAmount(), installmentNumber)));
+                    addToTransactionMapping(loanTransactionToRepaymentScheduleMapping, zero, zero, feeChargesPortion, zero);
+                }
+
+                loanTransaction.updateComponents(zero, zero, feeChargesPortion, penaltyChargesPortion);
+                unprocessed = loanTransaction.getAmount(currency).minus(paidAmount);
+                loanTransaction.updateLoanTransactionToRepaymentScheduleMappings(transactionMappings);
+            }
+        }
+
+        if (unprocessed.isGreaterThanZero()) {
+            handleOverAmount(loanTransaction, currency, installments, unprocessed, charges);
```

Comment:
In the handle over amount there is a piece of code which goes through the charges and checks whether any charge  was paid by the loan transaction. We need to exclude the amount that was already paid for the "original charges" and consider on the "overpay amount" might paid other charges.

Reply:
I think it would make sense to check what charge paid by relation is already on the loan transaction and the total amount of those are lower then the fee / penalty portion on the transaction. If so, we need to create the new charge paid by relations.

Also it would make sense to consider moving this whole logic to the point where we are distributing the amount of the loan transaction and if a fee / penalty got partially or totally paid, we create these relations immediately!